### ISSUE TITLE
macro expansion failure 'use-package

### ISSUE THREAD

**Issue by Luis-Henriquez-Perez** [raxod502/straight.el#203]
After I updated use-package to version 2.3 using `straight-pull-package`  on `use-package` I began getting this error. `(void-function use-package)` which is triggered when I first try to install a package with `(use-package highlight-indentation)`.

When I look in my messages buffer I see 

    Wrote ~/.emacs.d/D.el ...
    Loading /home/luis/.emacs.d/D.el (source)...
    Building straight...done
    Building use-package...
    Building use-package → Building bind-key...done
    Building use-package...done
    Eager macro-expansion failure: (void-function use-package) [2 times]
    Entering debugger...

I have this code which seems like it should be right. It was worked before I updated use-package.

     (setq package-enable-at-startup nil)
     (let ((bootstrap-file (concat user-emacs-directory "straight/bootstrap.el"))
           (bootstrap-version 2))
           (unless (file-exists-p bootstrap-file)
      (with-current-buffer
          (url-retrieve-synchronously
           "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
           'silent 'inhibit-cookies)
        (goto-char (point-max))
        (eval-print-last-sexp)))
        (load bootstrap-file nil 'nomessage))

    (straight-use-package 'use-package)
    (setq straight-use-package-by-default t) 
    (setq straight-enable-use-package-integration t)
    (setq use-package-always-defer t
        use-package-always-ensure t)

I've reviewed the changes in `use-package` concerning straight and have tried experimenting with using `:straight t` instead of defining `use-package-always-ensure t`. I've also tried rebuilding both `straight` and `use-package`. I haven't been able to figure this out. Any ideas of what I can do to keep debugging?

**Comment by Luis-Henriquez-Perez** [comment]
I don't know what exactly was wrong but It had to do with using the `:ensure t` word which is now not used with `straight`. When using `:ensure t` `use-package` assumes that the package manager is package.el.

**Comment by raxod502** [comment]
If you set both `straight-use-package-by-default` and `use-package-always-ensure` to non-nil values, then `use-package` will attempt to install packages using _both_ `straight.el` and `package.el`. This is almost certainly not what you would want, but I don't know why it would cause the error you described. In any case, I'm glad you resolved the error.

**Comment by tummychow** [comment]
I can also reproduce this bug. [This](https://github.com/tummychow/emacs/commit/cc5f495a659b2bcf194f7924940e139ac35ee5be) is my work-in-progress commit for the update. It touches use-package, straight, and general (which also has a use-package keyword).

Looking at my build directory gives me a hint of why the problem is occurring:

```
[~/.emacs.d/straight/build/use-package] [new-use-package|?2] $ pwd
/home/sjung/.emacs.d/straight/build/use-package
[~/.emacs.d/straight/build/use-package] [new-use-package|?2] $ ls
use-package-autoloads.el  use-package.el
```

Despite being on the new use-package version, none of the new files (use-package-core.el, etc) have been included in the build. At some point emacs tries to require `'use-package-core` and fails. I've tried straight-rebuild-package to no effect. Any ideas of what to try next?

For the record, I've also tried cloning my config (from my master branch, without this update) and bootstrapping straight completely from scratch. It fails immediately after building use-package, with the message `package-installed-p: package.el is not yet initialized!`.

**Comment by raxod502** [comment]
You probably forgot to update your `~/.emacs.d/straight/repos/melpa`, so the recipe for `use-package` was outdated. Running `M-x straight-get-recipe RET use-package RET` will show you what recipe `straight.el` is using for `use-package`.

Also, I looked at the [commit](https://github.com/tummychow/emacs/commit/cc5f495a659b2bcf194f7924940e139ac35ee5be) you linked, and I noticed that you updated `bootstrap-version` without actually updating the bootstrap snippet itself. That will cause your configuration to fail to bootstrap, since version 2 of the bootstrap snippet expects there to be a symlink at `~/.emacs.d/straight/bootstrap.el` which is not created when `bootstrap-version` is 3 or greater.
